danger_python/danger.py

```diff
@@ -58,31 +58,37 @@ def __init__(self):
         if not Danger.results:
             Danger.results = DangerResults()
 
-    dsl: DangerDSLJSONType = None
-    results: DangerResults = None
+    dsl: Optional[DangerDSLJSONType] = None
+    results: Optional[DangerResults] = None
 
     @property
-    def git(self) -> GitJSONDSL:
+    def git(self) -> Optional[GitJSONDSL]:
+        assert Danger.dsl is not None
         return Danger.dsl.git
 
     @property
-    def github(self) -> GitHubDSL:
+    def github(self) -> Optional[GitHubDSL]:
+        assert Danger.dsl is not None
         return Danger.dsl.github
 
     @property
-    def bitbucket_cloud(self) -> BitBucketCloudJSONDSL:
+    def bitbucket_cloud(self) -> Optional[BitBucketCloudJSONDSL]:
+        assert Danger.dsl is not None
         return Danger.dsl.bitbucket_cloud
 
     @property
-    def bitbucket_server(self) -> BitBucketServerJSONDSL:
+    def bitbucket_server(self) -> Optional[BitBucketServerJSONDSL]:
+        assert Danger.dsl is not None
         return Danger.dsl.bitbucket_server
 
     @property
-    def gitlab(self) -> GitLabDSL:
+    def gitlab(self) -> Optional[GitLabDSL]:
+        assert Danger.dsl is not None
         return Danger.dsl.gitlab
 
     @property
-    def settings(self) -> DangerDSLJSONTypeSettings:
+    def settings(self) -> Optional[DangerDSLJSONTypeSettings]:
+        assert Danger.dsl is not None
         return Danger.dsl.settings
```

Comment:
Are you sure you want all these as optional? In TypeScript I purposefully made them not optional (even though they are) because end-users will probably only use the right one in the right place. 

( I'm not sure if optionality is something you're _forced_ to deal with in python though )

Reply:
I think this should be fine because those are only type-hints and you do not deal with optionals in Python, hence it doesn't imply any changes in the `dangerfile.py`. You still do reference it as a non-optional type:

```python
title = danger.github.pr.title
markdown(title)
```

Python has no null checking built-in, it only uses the types as documentation and hints for the code completion. There's a tool called [mypy](https://mypy.readthedocs.io/en/stable/) which aims to check type hints statically and I made the change since the tool has detected the problem (and the point is valid, technically it is optional). 

There is also a marginal difference in what the end-user will see as the stack trace. Given the current implementation and the `dangerfile.py` contents defined as:

```python
title = danger.bitbucket_server.pr.title
markdown(title)
```

On the missing key, it prints out:

```
There was an error when executing dangerfile.py:
AttributeError at line 1: 'NoneType' object has no attribute 'pr'

Stacktrace:
  File "dangerfile.py", line 1, in <module>
    title = danger.bitbucket_server.pr.title
```

With the assertion in place it becomes:

```
There was an error when executing dangerfile.py:
AssertionError at line 1: 

Stacktrace:
  File "dangerfile.py", line 1, in <module>
    title = danger.bitbucket_server.pr.title
  File "/Users/kuba/dev/elp/projects/danger-python/danger_python/danger.py", line 77, in bitbucket_server
    assert server is not None
AssertionError
```

Both are very readable, so I think this can stay in the current form. What do you think?